pkg/controllers/deprovisioning/consolidation.go

```diff
@@ -0,0 +1,286 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deprovisioning
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter-core/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/controllers/provisioning"
+	"github.com/aws/karpenter-core/pkg/controllers/state"
+	"github.com/aws/karpenter-core/pkg/events"
+	"github.com/aws/karpenter-core/pkg/metrics"
+	"github.com/aws/karpenter-core/pkg/scheduling"
+	"github.com/aws/karpenter-core/pkg/utils/pod"
+)
+
+// Consolidation is the consolidation controller.
+type Consolidation struct {
+	kubeClient    client.Client
+	cluster       *state.Cluster
+	provisioner   *provisioning.Provisioner
+	recorder      events.Recorder
+	clock         clock.Clock
+	cloudProvider cloudprovider.CloudProvider
+}
+
+func NewConsolidation(clk clock.Clock, kubeClient client.Client, provisioner *provisioning.Provisioner,
+	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Consolidation {
+	return &Consolidation{
+		clock:         clk,
+		kubeClient:    kubeClient,
+		cluster:       cluster,
+		provisioner:   provisioner,
+		recorder:      recorder,
+		cloudProvider: cp,
+	}
+}
+
+// shouldNotBeDeprovisioned is a predicate used to filter deprovisionable nodes
+func (c *Consolidation) shouldNotBeDeprovisioned(_ context.Context, n *state.Node, provisioner *v1alpha5.Provisioner, _ []*v1.Pod) bool {
+	return provisioner == nil || provisioner.Spec.Consolidation == nil || !ptr.BoolValue(provisioner.Spec.Consolidation.Enabled) ||
+		n.Node.Annotations[v1alpha5.DoNotConsolidateNodeAnnotationKey] == "true"
+}
+
+// sortCandidates orders deprovisionable nodes by the disruptionCost
+func (c *Consolidation) sortCandidates(nodes []candidateNode) []candidateNode {
+	sort.Slice(nodes, func(i int, j int) bool {
+		return nodes[i].disruptionCost < nodes[j].disruptionCost
+	})
+	return nodes
+}
+
+// computeCommand generates a deprovisioning command given deprovisionable nodes
+//
+//nolint:gocyclo
+func (c *Consolidation) computeCommand(ctx context.Context, attempt int, candidates ...candidateNode) (deprovisioningCommand, error) {
+	// First delete any empty nodes we see
+	if attempt == 0 {
```

Comment:
If we have emptiness as a deprovisioner, why is this in consolidation as well?

Reply:
This was to achieve feature parity with Consolidation. Consolidation still deletes empty nodes even if emptiness is disabled. This will be eventually solved as an edge case of Consolidation in the future.